File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.rs

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let median = nums.len()/2;
        let n = nums[median];
        let mut node = TreeNode::new(n);
        
        let ls = nums[..median].to_vec();
        let rs = nums[median + 1..].to_vec();
        
        Self::merge(&mut node, ls);
        Self::merge(&mut node, rs);
        
        Self::wrap(node)
        
    }
    
    pub fn merge(node: &mut TreeNode,  nums: Vec<i32>){
        if (nums.len() <= 0){
            return;
        }
        
        let median_index = nums.len()/2;;
        
        let median = nums[median_index];
        
        let mut child = TreeNode::new(median);
        
        let ls = nums[..median_index].to_vec();
        let rs = nums[median_index + 1..].to_vec();
        
        Self::merge(&mut child, ls);
        Self::merge(&mut child, rs);
        
        let wrapped_child = Self::wrap(child);
        
        if median > node.val{
            node.right = wrapped_child
        }else{
            node.left = wrapped_child
        }
    }
    
    pub fn wrap(node: TreeNode) -> Option<Rc<RefCell<TreeNode>>>{
            Some(
                Rc::new(
                    RefCell::new(node)
                )
            )
    }
}

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution: 
    def bst(self, nums: List[int], start: int, end: int ) -> Optional[TreeNode]:
        l = end - start
        
        if end < start or l == 0:
            return None
        
        if l == 1:
            return TreeNode(nums[start])
        
        median_index = (int)(l/2) + start
        median = nums[median_index]
        node = TreeNode(median)
        
#         print("left: {start}, {end}".format(start = start, end = median_index))
#         print(nums[start:median_index])
        
#         print("right: {start}, {end}".format(start = median_index +1, end = end))
#         print(nums[median_index+1:l])
        node.left = self.bst(nums, start, median_index)
        node.right = self.bst(nums, median_index+1, end)
        
        return node
        
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        return self.bst(nums, 0, len(nums))
        